# Bioinformatics Specialization - UC San Diego / Coursera

---

NÂ°  | Status | Course |
--- | --- | --- |
I   | done | [Finding Hidden Messages in DNA](https://www.coursera.org/learn/dna-analysis)
II  | current | [Genome Sequencing](https://www.coursera.org/learn/genome-sequencing)
III | queued | [Comparing Genes, Proteins, and Genomes](https://www.coursera.org/learn/comparing-genomes) |
IV  | queued | [Molecular Evolution](https://www.coursera.org/learn/molecular-evolution) |
V   | queued | [Genomic Data Science and Clustering](https://www.coursera.org/learn/genomic-data) |
VI  | queued | [Finding Mutations in DNA and Proteins](https://www.coursera.org/learn/dna-mutations) |

<!--
VII | queued | [Bioinformatics Capstone: Big Data in Biology](https://www.coursera.org/learn/bioinformatics-project) |
-->

---
## I. Finding Hidden Messages in DNA

### 1. [Introduction](https://stepik.org/course/604)
+ Hidden message problem
+ Pattern matching problem
+ Clump finding problem

### 2. [Finding Replication Origins](https://stepik.org/course/605)
+ Frequent words problem
+ Approximate search and Hamming distance

### 3. [Hunting for Regulatory Motifs](https://stepik.org/course/606)
+ Greedy and pseudo-greedy motif search

### 4. [How Rolling Dice Helps Us Find Regulatory Motifs](https://stepik.org/course/607)
+ Randomized motif search and Gibbs sampling

---
## II. Genome Sequencing

### 1. [Introduction to Genome Sequencing](https://stepik.org/course/608)
+ String composition problem and genome assembly
+ Overlap graphs and Hamiltonian paths
+ De Bruijn Graphs and Eulerian Paths

### 2. [Applying Euler's Theorem to Assemble Genomes](https://stepik.org/course/609)
+ Read-Pairs and Virtual Reads
+ Assembling Genomes from Read Pairs
+ Dealing with Real Sequencing Data
  - imperfect coverage
  - error-prone reads and bubble detection/correction
  - inexact repeats
  - inferring k-mer multiplicities

### 3. [Sequencing Antibiotics](https://stepik.org/course/610)
+ Protein Translation
+ Mass Spectrometry and Cyclopeptide Sequencing

### 4. [From Ideal to Real Spectra for Antibiotics Sequencing](https://stepik.org/course/611)
+ Handling Noisy Spectra
+ Sequencing Peptides with Non-Proteinogenic Amino Acids
+ Spectral Convolutions

---
## III. Comparing Genes, Proteins, and Genomes

### 1. [Introduction to Sequence Alignment](https://stepik.org/course/612)
### 2. [From Finding a Longest Path to Aligning DNA Strings](https://stepik.org/course/613)
### 3. [Advanced Topics in Sequence Alignment](https://stepik.org/course/614)
### 4. [Genome Rearrangements and Fragility](https://stepik.org/course/615)
### 5. [Applying Genome Rearrangements Analysis to Find Genome Fragility](https://stepik.org/course/616)

---
## IV. Molecular Evolution

### 1. [Introduction to Evolutionary Tree Construction](https://stepik.org/course/617)
### 2. [More Algorithms for Constructing Trees from Distance Matrices](https://stepik.org/course/618)
### 3. [Constructing Evolutionary Trees from Characters](https://stepik.org/course/619)
### 4. [Was T. rex Just a Big Chicken?](https://stepik.org/course/620)
### 5. [Resolving the T. rex Peptides Mystery?](https://stepik.org/course/621)

---
## V. Genomic Data Science and Clustering

### 1. [Introduction to Clustering Algorithms](https://stepik.org/course/691)
### 2. [Advanced Clustering Techniques](https://stepik.org/course/692)
### 3. [Introductory Algorithms in Population Genetics](http://bioinformaticsalgorithms.com/excerpt/How%20Have%20Humans%20Populated%20the%20Earth.pdf)

---
## VI. Finding Mutations in DNA and Proteins

### 1. [Introduction to Read Mapping](https://stepik.org/course/686)
### 2. [The Burrows-Wheeler Transform](https://stepik.org/course/687)
### 3. [Speeding Up Burrows-Wheeler Read Mapping](https://stepik.org/course/688)
### 4. [Introduction to Hidden Markov Models](https://stepik.org/course/5273)
### 5. [Profile HMMs for Sequence Alignment](https://stepik.org/course/5274)
